internals/plan/plan.go

```diff
@@ -641,179 +646,219 @@ func CombineLayers(layers ...*Layer) (*Layer, error) {
 		}
 	}
 
-	// Ensure fields in combined layers validate correctly (and set defaults).
-	for name, service := range combined.Services {
-		if service.Command == "" {
-			return nil, &FormatError{
-				Message: fmt.Sprintf(`plan must define "command" for service %q`, name),
-			}
+	// Set defaults where required.
+	for _, service := range combined.Services {
+		if !service.BackoffDelay.IsSet {
+			service.BackoffDelay.Value = defaultBackoffDelay
+		}
+		if !service.BackoffFactor.IsSet {
+			service.BackoffFactor.Value = defaultBackoffFactor
+		}
+		if !service.BackoffLimit.IsSet {
+			service.BackoffLimit.Value = defaultBackoffLimit
+		}
+	}
+
+	for _, check := range combined.Checks {
+		if !check.Period.IsSet {
+			check.Period.Value = defaultCheckPeriod
 		}
+		if !check.Timeout.IsSet {
+			check.Timeout.Value = defaultCheckTimeout
+		}
+		if check.Threshold == 0 {
+			// Default number of failures in a row before check triggers
+			// action, default is >1 to avoid flapping due to glitches. For
+			// what it's worth, Kubernetes probes uses a default of 3 too.
+			check.Threshold = defaultCheckThreshold
+		}
+	}
+
+	return combined, nil
+}
+
+// Validate checks that the layer is valid. It returns nil if all the checks pass, or
+// an error if there are validation errors.
+// See also Plan.Validate, which does additional checks based on the combined
+// layers.
+func (layer *Layer) Validate() error {
```

Comment:
> It looks like you've demoted some post-combine checks here back to layer-time checks. I like the idea of failing on an issue as soon as possible, since the error context will also be more accurate.

Yes, that was my intention with this, and to logically separate out the code that's checking that a `Layer` is valid from checking that a combination of `Layer`s is valid.

> How does this block of validation compare with what is in `ParseLayer` ? All new layers start their life in `ParseLayer` (apart from `SetServiceArgs` I believe), and some layer validation already appears present there.

Indeed; nice catch. I don't think there's any overlap but the checks are fairly similar in nature (`ParseLayer` being mostly around the names).

@benhoyt what's your opinion on moving out the validation pieces of `ParseLayer` into `Layer.Validate`? And should `SetServiceArgs` do a `newLayer.Validate()`?

Reply:
Yeah, good point. I do think we should move all the layer-validation logic from `ParseLayer` into `Layer.Validate`, and then just call `layer.Validate()` inside `ParseLayer`. Much tidier.

Should `SetServiceArgs` validate? It probably wouldn't need to as we're building that layer from scratch so it shouldn't have error, but it also wouldn't hurt, and would be quite cheap -- so coding defensively I think we might as well.